Add 'ensuring releases perform at scale' doc (#7409)
mattermost/docs: source/scale/ensuring-releases-perform-at-scale.rst

Ensuring Releases Perform at Scale

> ======================

Reviewer: Please ensure that all heading indicators on the page equal the length of the title text (to avoid build warnings/errors).

> To ensure each release of Mattermost upholds our high standards for performance at scale the team performs thorough load testing, develops features with scale in mind, and follows strict guidelines for database schema migrations.

Reviewer: ```suggestion
To ensure each release of Mattermost upholds our high standards for performance at scale, the Mattermost Engineering team performs thorough load testing, develops features with scale in mind, and follows strict guidelines for database schema migrations.
```

Monthly release load tests

> ---------------------------------

Reviewer: FYI: Longer heading indicators are better than ones that are too short :)

Each month, before being approved for distribution, a release candidate of Mattermost is load tested via a comprehensive and mature set of `load test tooling <https://github.com/mattermost/mattermost-load-test-ng>`__ using simulated data that matches real-world, high-scale usage patterns.

> Multiple tests covering different configurations are run against highly-available deployments of Mattermost with thousands of users and millions of posts. Both PostgreSQL and MySQL are tested, although MySQL tests will be dropped when :doc:`MySQL goes out of support with Mattermost v11 </about/deprecated-features#mattermost-server-v10-0-0>`.

Reviewer: ```suggestion
Multiple tests covering different configurations are run against highly-available deployments of Mattermost with thousands of users and millions of posts. Both PostgreSQL and MySQL are tested, although MySQL tests will be dropped when :ref:`MySQL goes out of support with Mattermost v11 <about/deprecated-features:mattermost server v10.0.0>`.
```

> The load tests generate a report detailing average API request times, database i/o, memory usage, concurrency, requests per second and more. This report on the performance of the release candidate is then compared to the previous month's report. Any non-normal deviations are investigated and remedied before the release candidate is promoted to the final release.

Reviewer 2: non-normal sounded a bit off, maybe abnormal? 0/5

> The load tests generate a report detailing average API request times, database i/o, memory usage, concurrency, requests per second and more. This report on the performance of the release candidate is then compared to the previous month's report. Any non-normal deviations are investigated and remedied before the release candidate is promoted to the final release.

Reviewer: ```suggestion
The load tests generate a report detailing average API request times, database I/O, memory usage, concurrency, requests per second, and more. This performance report on the release candidate is then compared to the previous month's report. Any deviations are investigated and remedied before the release candidate is promoted to the final release.
```

> The load tests generate a report detailing average API request times, database I/O, memory usage, concurrency, requests per second, and more. This performance report on the release candidate is then compared to the previous month's report. Any deviations are investigated and remedied before the release candidate is promoted to the final release.

Reviewer 3: > This performance report on the release candidate is then compared to the previous month's report.

This is slightly misleading: we don't compare against previous reports, but against the latest version. That is: we run a test and generate a report on the latest stable version, and the same test and generate another report on the latest RC version to be released; e.g., `v9.11.0` vs `v10.0.0-rc2` (https://github.com/mattermost/performance-reports/tree/main/performance-comparisons/v10.0). Then we compare those two reports. The analysis that we publish (like the one I just linked is then on the comparison between those two reports.

Author: Thanks! That's good clarification. How's the update look?

Reviewer 3: That's perfect! Thanks :)

> Each report, along with analysis, is posted into the public `~Developers: Performance <https://community.mattermost.com/core/channels/developers-performance>`__ channel.

Reviewer: ```suggestion
Each report, along with analysis, is posted into the public `Developers: Performance <https://community.mattermost.com/core/channels/developers-performance>`__ channel.
```

Developing scalable features and systems
--------------------------------

> Scale is a major consideration during the development of new features and systems. It is included in technical specifications from the beginning of the software design process.

Reviewer: ```suggestion
Scale is a major consideration during the development of new features and systems, and it's included in technical specifications from the beginning of the software design process.
```

As a part of implementation, the load test agent is updated to include coverage for the feature or system. A load test is then run with results being compared to baselines.

The code changes are then reviewed by at least two developers and a SDET/QA analyst before being merged.

> Once merged, a new build is created and deployed the next day to `community.mattermost.com <https://community.mattermost.com>`__ where any impacts on performance are monitored for 3-4 weeks before the changes are included in a release candidate.

Reviewer: ```suggestion
Once merged, a new build is created and deployed the next day to the `Mattermost Community Server <https://community.mattermost.com>`__ where any impacts on performance are monitored for 3-4 weeks before the changes are included in a release candidate.
```

Database schema changes
-----------------------------

Database schema changes are kept to a minimum to reduce risk on upgrades and impact on performance. When a change in schema is required, Mattermost follows `strict migration guidelines <https://developers.mattermost.com/contribute/more-info/server/schema-migration-guide/>`__ that minimize risk and prevents performance impact during and after migrations.

If a more involved migration is required, detailed analysis is performed and published with guidance. An example analysis is the `Mattermost v6.0 schema migration <https://gist.github.com/streamer45/59b3582118913d4fc5e8ff81ea78b055>`__.

Additionally, all database schema changes are load tested as a part of our monthly release process.

Monitoring post-release
-----------------------------

After a new version of Mattermost is released it is rolled out to Mattermost Cloud customers over a multi-week period. During rollout performance and error rate metrics are monitored for any unexpected changes. If any user-impacting changes are observed, the release is reverted and the deviations are investigated with any fixes being delivered as a part of patch release.

> Additionally Mattermost channels, user forums and support tickets are closely monitored for reports of any issues. Any reports are investigated and resolved appropriately.

Reviewer: ```suggestion
In addition, Mattermost channels, user forums, and support tickets are closely monitored for reports of any issues. Any reports are investigated and resolved appropriately.
```
